picard/config.py

```diff
@@ -341,12 +322,51 @@ class ListOption(Option):
 setting = None
 persist = None
 
+PURGE_INTERVAL_MILLISECONDS = 60000
+_thread_configs = {}
+_thread_config_lock = threading.RLock()
+
 
-def _setup(app, filename=None):
+def setup_config(app, filename=None):
     global config, setting, persist
     if filename is None:
         config = Config.from_app(app)
     else:
         config = Config.from_file(app, filename)
+    _thread_configs[threading.get_ident()] = config
     setting = config.setting
     persist = config.persist
+    QtCore.QTimer.singleShot(PURGE_INTERVAL_MILLISECONDS, _purge_config_instances)
+
+
+def get_config():
+    """Returns a config object for the current thread.
+
+    Config objects for threads are created on demand and cached for later use.
+    """
+    thread_id = threading.get_ident()
+    thread_config = _thread_configs.get(thread_id)
+    if not thread_config:
+        _thread_config_lock.acquire()
+        try:
+            config_file = config.fileName()
+            log.debug('Instantiating Config for thread %s using %s.', thread_id, config_file)
+            thread_config = Config.from_file(None, config_file)
+            _thread_configs[thread_id] = thread_config
+        finally:
+            _thread_config_lock.release()
+    return thread_config
+
+
+def _purge_config_instances():
+    """Removes cached config instances for no longer active threads."""
+    _thread_config_lock.acquire()
+    try:
+        all_threads = set([thread.ident for thread in threading.enumerate()])
+        threads_config = set(_thread_configs)
+        for thread_id in threads_config.difference(all_threads):
+            log.debug('Purging config instance for thread %s.', thread_id)
+            del _thread_configs[thread_id]
+    finally:
+        _thread_config_lock.release()
+        QtCore.QTimer.singleShot(PURGE_INTERVAL_MILLISECONDS, _purge_config_instances)
```

Comment:
What about something like:

```python
def _purge_configs_timer_start(purge_interval_milliseconds=60000):
    def _purge_config_instances():
        """Removes cached config instances for no longer active threads."""
        _thread_config_lock.acquire()
        try:
            all_threads = set([thread.ident for thread in threading.enumerate()])
            threads_config = set(_thread_configs)
            for thread_id in threads_config.difference(all_threads):
                log.debug('Purging config instance for thread %s.', thread_id)
                del _thread_configs[thread_id]
        finally:
            _thread_config_lock.release()
            _purge_configs_timer_start(purge_interval_milliseconds)

    QtCore.QTimer.singleShot(purge_interval_milliseconds, _purge_config_instances)
```

Reply:
Instead of defining a new `_purge_config_instances` per call I'd rather put this on module level, at which point `_purge_configs_timer_start` is just a call to `QTimer.singleShot`. Maybe if you want to have the interval encapsulated something like:

```
def _init_purge_config_timer(purge_interval_milliseconds=60000):
    def purge_config_instances():
        """Removes cached config instances for no longer active threads."""
        _thread_config_lock.acquire()
        try:
            all_threads = set([thread.ident for thread in threading.enumerate()])
            threads_config = set(_thread_configs)
            for thread_id in threads_config.difference(all_threads):
                log.debug('Purging config instance for thread %s.', thread_id)
                del _thread_configs[thread_id]
        finally:
            _thread_config_lock.release()
            start_purge_config_timer()

    def start_purge_config_timer():
        QtCore.QTimer.singleShot(purge_interval_milliseconds, purge_config_instances)

    start_purge_config_timer()
```

Reply:
What do you think about 8053e59 ?